test_semver.py

```diff
@@ -580,3 +584,44 @@ def test_should_be_able_to_use_integers_as_prerelease_build():
     assert isinstance(v.prerelease, str)
     assert isinstance(v.build, str)
     assert VersionInfo(1, 2, 3, 4, 5) == VersionInfo(1, 2, 3, '4', '5')
+
+
+@pytest.mark.parametrize("version,parts,expected", [
+    ("3.4.5", dict(major=2), '2.4.5'),
+    ("3.4.5", dict(major="2"), '2.4.5'),
+    ("3.4.5", dict(major=2, minor=5), '2.5.5'),
+    ("3.4.5", dict(minor=2), '3.2.5'),
+    ("3.4.5", dict(major=2, minor=5, patch=10), '2.5.10'),
+    ("3.4.5", dict(major=2, minor=5, patch=10, prerelease="rc1"),
+     '2.5.10-rc1'),
+    ("3.4.5", dict(major=2, minor=5, patch=10, prerelease="rc1", build="b1"),
+     '2.5.10-rc1+b1'),
+    ("3.4.5-alpha.1.2", dict(major=2), '2.4.5-alpha.1.2'),
+    ("3.4.5-alpha.1.2", dict(build="x1"), '3.4.5-alpha.1.2+x1'),
+    ("3.4.5+build1", dict(major=2), '2.4.5+build1'),
+])
+def test_should_replace(version, parts, expected):
+    assert replace(version, **parts) == expected
+
+
+def test_should_raise_typeerror_for_invalid_replace():
+    with pytest.raises(KeyError):
+        assert replace("1.2.3", unknown="should_raise")
+
+
+@pytest.mark.parametrize("version,parts,expected", [
+    ("3.4.5", dict(major=2, minor=5), '2.5.5'),
+    ("3.4.5", dict(major=2, minor=5, patch=10), '2.5.10'),
+    ("3.4.5-alpha.1.2", dict(major=2), '2.4.5-alpha.1.2'),
+    ("3.4.5-alpha.1.2", dict(build="x1"), '3.4.5-alpha.1.2+x1'),
+    ("3.4.5+build1", dict(major=2), '2.4.5+build1'),
+])
+def test_should_versioninfo_replace(version, parts, expected):
+    v1 = VersionInfo.parse(version)
+    assert v1.replace(**parts) == VersionInfo.parse(expected)
+
+
+def test_should_raise_value_error_for_versioninfo_replace():
+    with pytest.raises(KeyError):
+        v1 = VersionInfo.parse("1.2.3")
+        v1.replace(major="x")
```

Comment:
A key error is wrong here. It should raise a ValueError. Good test case for that (ValueError instead of KeyError) though!

Reply:
Actually that was on purpose to be consistent. My first implementation was different: Depending on which function you call (`semver.replace` or semver.VersionInfo.replace`) you would get different exceptions (ValueError vs. TypeError, if I remember correctly).

Unfortunately, the `bump_*` functions are inconsistent what they raise. So...

Reply:
I'd rather be correct for new functionality and fix the old later. This ought to be a ValueError/TypeError as even the test name states

Why are the `bump_*` function inconsistent? They only take a version so only error is a ValueError. Using "x" for a version (part) is a ValueError. So KeyError is wrong here. Or am I missing anything?

Reply:
> Why are the bump_* function inconsistent?

Hmn... I can't reproduce it anymore. Probably it was due to that we coerced the major, minor, and patch arguments to integers which changed the exceptions too... :thinking: 

Anyway, yes, I agree, in that case it would be better to check for a ValueError exception. That would avoid the try...except handling after line 302; it would be unnecessary.

Reply:
I've corrected this.